accounts/forms.py

```diff
@@ -14,6 +20,122 @@ class Meta:
+            "Please enter a correct %(username)s and password. Note that email "
+            "and password fields may be case-sensitive. "
+            "If you have turned on two-factor authentication, you will need "
+            "an OTP authorization code from Google Authenticator as well. "
+        ),
+        "inactive": _("This account is inactive."),
+    }
+
+    auth_code = TOTPField(  # Do not require otp code in login views.
+        label="Two Factor Auth Code", help_text="6 digit one-time code", required=False
+    )
+
+    def clean(self):
+        """Hook that gets run after all the fields are cleaned.
+
+        Note: we are technically breaking LSP here since this is changing the
+            functionality of the superclass rather than extending it, but
+            AuthenticationForm is so dang close to what we want that it's ok
+            to bend the rules a little bit.
+        """
+        username = self.cleaned_data.get("username")
+        password = self.cleaned_data.get("password")
+
+        # Auth code is "Optional" in the sense that TwoFactorAuthenticationBackend
+        # will just flip a `using_2FA` flag as a signal to later requests - without
+        # this flag, Experimenters will be blocked from experimenter views.
+        auth_code = self.cleaned_data.get("auth_code")
+
+        if username is not None and password:
+            # Don't need ObjectPermissionBackend - just cut to the chase.
+            self.user_cache = two_factor_auth_backend.authenticate(
+                self.request, username=username, password=password, auth_code=auth_code
+            )
+
+            if self.user_cache is None:
+                raise self.get_invalid_login_error()
+            else:
+                # Set the backend path as django.contrib.auth.authenticate would usually,
+                # django.contrib.auth.login needs to know. This MUST also match the
+                # class name listed in settings.AUTHENTICATION_BACKENDS in order for
+                # AuthenticationMiddleware's request processor to work correctly, see:
+                # https://github.com/django/django/blob/master/django/contrib/auth/__init__.py#L179
+                self.user_cache.backend = (
+                    "accounts.backends.TwoFactorAuthenticationBackend"
+                )
+                self.confirm_login_allowed(self.user_cache)
+
+        return self.cleaned_data
+
+
+class ResearcherRegistrationForm(UserCreationForm):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # Don't autofill passwords, in the interest of security.
+        self.fields["password1"].widget.attrs["autocomplete"] = "new-password"
+        self.fields["password2"].widget.attrs["autocomplete"] = "new-password"
+
+    def save(self, commit=True):
+        """Just flip the active and researcher flags."""
+        user = super().save(commit=False)
+        user.is_active = True
+        user.is_researcher = True
```

Comment:
should probably move the post-save signal `add_researcher_to_labs` from studies/models.py here

Reply:
just bumping this up - does it make sense to do this now or let it be?

Reply:
IMO the lab member-adding logic is fine where it is - is there a particular reason you'd want to bring it into the save method?

Reply:
I was under the impression you didn't like putting logic into the signals if we didn't have to, but it works fine there - I'm happy with it if you are!